crates/runtime/src/embeddings/array_distance.rs

```diff
@@ -0,0 +1,210 @@
+            DataType::new_fixed_size_list(DataType::Float32, FIXED_SIZE_LIST_WILDCARD, false),
+            DataType::new_fixed_size_list(DataType::Float32, FIXED_SIZE_LIST_WILDCARD, true),
+        ];
+
+        Self {
+            signature: Signature::uniform(2, valid_types, Volatility::Immutable),
+        }
+    }
+}
+
+impl ScalarUDFImpl for ArrayDistance {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+    fn name(&self) -> &str {
+        "array_distance"
+    }
+    fn signature(&self) -> &Signature {
+        &self.signature
+    }
+
+    /// [`ArrayDistance`] expects two arguments of type `FixedSizeList<Float32>`. The two
+    /// arguments must have the same size, but may be any size together.
+    fn return_type(&self, args: &[DataType]) -> DataFusionResult<DataType> {
+        if args.len() != 2 {
+            return plan_err!("array_distance takes exactly two arguments");
+        }
+
+        match (args[0].clone(), args[1].clone()) {
+            (DataType::FixedSizeList(f1, size1), DataType::FixedSizeList(f2, size2)) => {
+                if f1.data_type() != &DataType::Float32 {
+                    return plan_err!("array_distance requires first arguments to be of type FixedSizeList<Float32>");
+                } else if f2.data_type() != &DataType::Float32 {
+                    return plan_err!("array_distance requires second arguments to be of type FixedSizeList<Float32>");
+                }
+                if size1 != size2 {
+                    return plan_err!(
+                        "array_distance requires both arguments to be of the same size"
+                    );
+                }
+
+                Ok(DataType::Float32)
+            }
+            (DataType::FixedSizeList(_f1, _size1), _) => {
+                return plan_err!(
+                    "array_distance requires the second argument to be of type FixedSizeList"
+                )
+            }
+            (_, DataType::FixedSizeList(_f1, _size1)) => {
+                return plan_err!(
+                    "array_distance requires the first argument to be of type FixedSizeList"
+                )
+            }
+            _ => {
+                return plan_err!(
+                    "array_distance requires the first argument to be of type FixedSizeList"
+                )
+            }
+        }
+    }
+
+    // Basic implementation of the Euclidean distance between two arrays
+    fn invoke(&self, args: &[ColumnarValue]) -> DataFusionResult<ColumnarValue> {
+        let args = ColumnarValue::values_to_arrays(args)?;
+
+        let v1: Vec<Float32Array> = as_fixed_size_list_array(&args[0])?
+            .iter()
+            .map(|v| {
+                v.ok_or_else(|| DataFusionError::Internal("no null entries allowed".into()))
+                    .and_then(|vv| {
+                        Arc::clone(&vv)
+                            .as_any()
+                            .downcast_ref::<Float32Array>()
```

Comment:
This `.downcast_ref::<Float32Array>()` is failing for reasons I don't understand.